Improve performance of encoding

taiki-e opened the pull request:
This change makes *encoding* about 7x faster than the current implementation, on my machine.

Before:

```
hex_encode              time:   [77.149 us 77.581 us 78.015 us]
```

After:

```
hex_encode              time:   [10.214 us 10.270 us 10.336 us]                        
```


<details>

<summary> Benchmarks per commit</summary>

Current main branch (aa8f3001fed80718ea8e33ab5e278351d3ef5036)

```
hex_encode              time:   [77.149 us 77.581 us 78.015 us]
```

First commit of this PR: Adjust #[inline] on encoding functions (97abf032d3430af4b2f6fa9b85806d99b2585b9d)

```
hex_encode              time:   [37.251 us 37.412 us 37.583 us]                        
                        change: [-52.047% -51.748% -51.412%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Second commit of this PR: Use encode_to_slice in encode (0129182ab8c57f9962c52367dc9665d69a3acab2)

```
hex_encode              time:   [27.029 us 27.230 us 27.476 us]                        
                        change: [-27.711% -27.244% -26.782%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Third commit of this PR: Use chunks_exact_mut instead of generate_iter (19b333d91cbbc58274cdcd948ef0b2f6078b5a2b)

```
hex_encode              time:   [10.214 us 10.270 us 10.336 us]                        
                        change: [-62.198% -61.839% -61.444%] (p = 0.00 < 0.05)
                        Performance has improved.
```


</details>

This also adds encode_to_slice_upper, which is needed to improve the performance of encode_upper. (fixes #45).

taiki-e committed:
Use lookup table for decoding

taiki-e committed:
Inline val function

taiki-e committed:
Use chunks_exact instead of chunks

taiki-e committed:
Use decode_to_slice in Vec::from_hex

taiki-e committed:
Inline decode_to_slice function

taiki-e committed:
Fix build errors

taiki-e committed:
Adjust #[inline] on encoding functions

taiki-e committed:
Use encode_to_slice in encode

taiki-e committed:
Use chunks_exact_mut instead of generate_iter

taiki-e commented on src/lib.rs:
```diff
@@ -257,7 +257,10 @@ from_hex_array_impl! {
+    let data = data.as_ref();
+    let mut out = vec![0; data.len() * 2];
+    encode_to_slice(data, &mut out).unwrap();
+    String::from_utf8(out).unwrap()
```
When using from_utf8_unckecked here, it can improve performance by about 8%. (It is safe because we emit only hex characters.)
However, I didn't apply that change because I don't know the policy regarding the unsafe code in this crate.
If It is okay with using unsafe code, I'll add that change.

```
hex_encode              time:   [9.4555 us 9.4828 us 9.5106 us]                        
                        change: [-9.2447% -8.3200% -7.4582%] (p = 0.00 < 0.05)
                        Performance has improved.
```

cgwalters replied:
Some overlap with https://github.com/KokaKiwi/rust-hex/pull/66 here I think.  In particular I suspect the combination of `ExactSizeIterator` will allow us to collect into a `String` efficiently without indirecting through `Vec`.

taiki-e committed:
Add encode_to_slice_upper to improve performance of encode_upper

This also fixes #45.

yjhmelody commented:
Nice PR! But the owner of the repo seems to be inactive.

KokaKiwi committed:
Merge branch 'main' into pref

KokaKiwi committed:
Merge branch 'pref' of github.com:taiki-e/rust-hex into taiki-e-pref

KokaKiwi committed:
refactor: improve code style and clarity in src/lib.rs

- Remove unused `u8` import from `core` module
- Add lint allow annotation for `__` constant to suppress non_upper_case_globals warning
- Refactor `decode_in_slice()` to use slice-based `val()` function for cleaner implementation

KokaKiwi committed:
Apply suggestions from code review

Co-authored-by: Lucas <24826124+Luro02@users.noreply.github.com>

KokaKiwi committed:
Merge branch 'pref' of github.com:taiki-e/rust-hex into taiki-e-pref

KokaKiwi committed:
Merge pull request #62 from taiki-e/pref

KokaKiwi committed:
Merge branch 'perf-encode' of github.com:taiki-e/rust-hex into taiki-e-perf-encode

KokaKiwi committed:
chore: fixup warning
